Release pipeline

sjorsholtrop-ritense opened the pull request:
Adds `release.yml` workflow to release new versions of Valtimo backend and frontend in lock-step.

Workflow must be run manually from a `release/*` branch. 

The workflow will build, test and release the frontend and backend libraries. Then it will build the frontend/backend images based on the GZAC implementation template in this repo, and push them to GHCR and Dockerhub (`ritense/gzac-backend`, `ritense/gzac-frontend`).

Secrets for Dockerhub login have been added to the repo.

sjorsholtrop-ritense committed:
add umbrella release piipeline

sjorsholtrop-ritense committed:
formatting and comments

sjorsholtrop-ritense committed:
use non-deprecated softprops/action-gh-release@v1

sjorsholtrop-ritense committed:
cicd: add dockerhub as image push targets, add ARM builds back

sjorsholtrop-ritense committed:
cicd: add dockerhub as push target for backend

ThomasMinkeRitense commented:
@sjorsholtrop-ritense two remarks:
- I'd like the version to be derived from the branch name. If I release from branch release/13.6.0, version should just be 13.6.0
- We can't use the publish to Docker Hub yet without cleaning up the configuration files and the test classes that are in the app module. I'd like to discuss this step with the team as well because I heard some concerns about publishing the development app to be our production worthy product.

sjorsholtrop-ritense committed:
infer release version from branch

sjorsholtrop-ritense commented:
Still blocked on cleaning up the frontend app by making it not include dev/test-only routes and components when building for production.

sjorsholtrop-ritense committed:
Stricter branch validation, add separate release candidate publishing workflows

sjorsholtrop-ritense committed:
fix caching logic, normalize branchnames for docker tags

sjorsholtrop-ritense committed:
add temporary PR trigger for testing

sjorsholtrop-ritense committed:
temp: pull_request_target over pull_request

sjorsholtrop-ritense committed:
fix determine_version

sjorsholtrop-ritense committed:
fix workflow dependencies

sjorsholtrop-ritense committed:
add RC testing

sjorsholtrop-ritense committed:
change workflow logic temporarily to test

sjorsholtrop-ritense committed:
change test trigger to push

sjorsholtrop-ritense committed:
fix workflow permissions

sjorsholtrop-ritense committed:
fix workflow permissions

sjorsholtrop-ritense committed:
fix versioning

sjorsholtrop-ritense committed:
fix workflow job names, add inherited secrets to frontend publish_image workflow call

sjorsholtrop-ritense commented:
Seems to work now. Tested it here: https://github.com/valtimo-platform/valtimo/actions/runs/20620909414

Publishing the frontend went wrong because I cannot publish over an existing 0.0.0 release, which was released in an earlier attempt.
Publishing the backend didn't work because I dropped the deployment. 
But publishing the image worked fine for both: `ritense/gzac-backend:0.0.0` and `ritense/gzac-frontend:0.0.0`. We can test those images now.


The last step in the pipeline, creating the GitHub Release, hasn't run yet. But I'm sure it'll work as it's just:
```
- name: Create release
  uses: softprops/action-gh-release@v1
  with:
    tag_name: ${{ needs.prepare.outputs.version }}
    name: ${{ needs.prepare.outputs.version }}
    body: "Release for Valtimo version ${{ needs.prepare.outputs.version }}."
    draft: false
    prerelease: false
```

sjorsholtrop-ritense committed:
make environment.prod.ts resemble gzac-frontend-template

mbritense committed:
Merge branch 'next-minor' into cicd/release-pipeline

mbritense committed:
Merge branch 'next-minor' into cicd/release-pipeline

mbritense commented:
@sjorsholtrop-ritense in `environment.prod.ts` production is false, but it should be true. I'm guessing this might cause the form builder issue in the production builds on the test environment

sjorsholtrop-ritense committed:
Merge branch 'next-minor' into cicd/release-pipeline

sjorsholtrop-ritense committed:
set production: true, build libs with --configuration production explicitly

sjorsholtrop-ritense committed:
fix tagging

sjorsholtrop-ritense committed:
merge next-minor

sjorsholtrop-ritense committed:
cicd: fix wrong use of github.env

sjorsholtrop-ritense commented on .github/workflows/frontend_publish_release.yml:
```diff
@@ -2,48 +2,50 @@ name: Frontend - Publish release
-    if: ${{ startsWith(github.ref_name, 'release/') }}
+  publish_libs:
     needs:
       - build_and_test_libs
-      - determine_version
     uses: ./.github/workflows/frontend_publish_to_npm.yml
     with:
-      project_version: ${{ needs.determine_version.outputs.project_version }}
+      project_version: ${{ github.env.PROJECT_VERSION }}
+      frontend_libraries_build_artifact_name: ${{ needs.build_and_test_libs.outputs.build_artifact_name }}
```
Good issue, fixed

sjorsholtrop-ritense committed:
fix softprops/action-gh-release version, set to v2

sjorsholtrop-ritense commented on .github/workflows/publish_release_candidate.yml:
```diff
@@ -0,0 +1,72 @@
+      - prepare
+    runs-on: ubuntu-latest
+    steps:
+      - name: Create release
+        uses: softprops/action-gh-release@v1
+        with:
+          tag_name: ${{ needs.prepare.outputs.version }}
+          name: ${{ needs.prepare.outputs.version }}
+          body: "Release Candidate for Valtimo version ${{ needs.prepare.outputs.version }}."
+          draft: false
+          prerelease: true
```
Good issue, fixed

sjorsholtrop-ritense commented on .github/workflows/publish_release.yml:
```diff
@@ -0,0 +1,73 @@
+      project_version: ${{ needs.prepare.outputs.version }}
+    secrets: inherit
+
+  create_release:
+    name: Create GitHub release
+    needs:
+      - backend_release
+      - frontend_release
+      - prepare
+    runs-on: ubuntu-latest
+    steps:
+      - name: Create release
+        uses: softprops/action-gh-release@v1
+        with:
+          tag_name: ${{ needs.prepare.outputs.version }}
+          name: ${{ needs.prepare.outputs.version }}
+          body: "Release for Valtimo version ${{ needs.prepare.outputs.version }}."
+          draft: false
+          prerelease: false
```
Good issue, fixed

sjorsholtrop-ritense committed:
remove temp on:push trigger

sjorsholtrop-ritense commented on .github/workflows/publish_release_candidate.yml:
```diff
@@ -0,0 +1,72 @@
+  workflow_dispatch:
+  push:
+    branches:
+      - rc/0.0.0 # Temporarily, for testing without having to have this workflow in the default branch
```
Fixed

sjorsholtrop-ritense committed:
Merge branch 'next-minor' into cicd/release-pipeline

sjorsholtrop-ritense committed:
Merge branch 'next-minor' into cicd/release-pipeline

sjorsholtrop-ritense committed:
make config similar to gzac-frontend-template

sjorsholtrop-ritense committed:
add smtp, documentgenerator, mailtemplate, texttemplate plugins

sjorsholtrop-ritense committed:
add SMTP, freemarker plugins to monorepo backend

sjorsholtrop-ritense committed:
Merge remote-tracking branch 'origin/next-minor' into cicd/release-pipeline

sjorsholtrop-ritense committed:
formatting reverts

sjorsholtrop-ritense committed:
update plugin versions to match v/13 branch

sjorsholtrop-ritense committed:
Merge branch 'next-minor' into cicd/release-pipeline

sjorsholtrop-ritense commented:
Left to do:

- [x] Must have: Fix the backend image, which does not work since adding the standard plugins in [5b20290](https://github.com/valtimo-platform/valtimo/pull/145/commits/5b20290cc887aefe31062dd318d2c0ec0ba720eb). 
Try a test-env (but don't keep the test env active for long since it crash loops)
- [ ] Must have: Do functional tests to see if the images built from the monorepo are identical to the ones from the templates
- [ ] Nice to have: In the pipeline, on successful release, update gradle.properties with the new version on `next-minor` / `next-major`
- [ ] Nice to have: In the pipeline, on successful release, merge the released branch to the v12-stable or v13-stable branch (see [issue](https://github.com/generiekzaakafhandelcomponent/atlas-internal/issues/140))
- [ ] Nice to have: For releasing patch versions of older releases, make sure the merge into `v12-stable` or `v13-stable` is optional.

ThomasMinkeRitense committed:
Merge next-minor

Klaas-Ritense committed:
Merge remote-tracking branch 'origin/next-minor' into cicd/release-pipeline

# Conflicts:
#	.github/workflows/frontend_build_push_docker_image.yml
#	frontend/Dockerfile
#	frontend/package.json

Klaas-Ritense committed:
cicd: bump plugin pins, fix release workflow blockers, add v12-stable opt-out

- frontend/Dockerfile: revert COPY path to deployment/valtimo-angular-console/browser
  to match what angular.json's application builder actually produces.
- backend_publish_release.yml: replace dangling steps.version.outputs.version_release
  with ${PROJECT_VERSION} in the JReleaser deploy step. Drop build_and_push_image
  from publish.needs: so libs publish independently of image build.
- publish_release.yml / publish_release_candidate.yml / backend+frontend
  publish_release.yml + _candidate.yml: add publish_images boolean input
  (default true), plumbed through the umbrella -> per-side workflows. Gate the
  image-build sub-workflow on it. Lets v12-stable opt out of the in-CI image
  build while v12 templates are still Jenkins-built.
- gradle.properties: freemarkerPluginVersion 8.2.0 -> 8.3.6,
  smtpmailPluginVersion 2.0.0 -> 2.0.4.
- frontend/package.json: @valtimo-plugins/freemarker 8.1.1 -> 8.3.6,
  @valtimo-plugins/smtpmail 2.0.1 -> 2.0.4.

Klaas-Ritense committed:
cicd: restore frontend_ci.yml artifact-name input lost in next-minor merge

The merge of next-minor reverted the 'with:' block that passes
frontend_libraries_build_artifact_name into frontend_build_push_docker_image.yml.
PR #145 made that input required, so the missing 'with:' caused 'Run frontend CI'
to startup_failure. Restoring the input.

Klaas-Ritense committed:
cicd: drop frontend smtpmail to 2.0.2 (Angular 19 compatible)

@valtimo-plugins/smtpmail@2.0.4's npm peerDep declares @angular/core ^17.2.2.
The project is on Angular 19, so the install left @valtimo/* workspace
symlinks broken and 'ng build' couldn't resolve them. 2.0.2 is the last
Angular-19-compatible release. Backend Maven pin stays at 2.0.4 (independent
ecosystem).

Klaas-Ritense committed:
cicd: restore lib-dist download + file: rewrite in frontend image build

esbuild can't resolve @valtimo/* imports that come from the installed
@valtimo-plugins/* tarballs because their .mjs files reference workspace
packages that aren't installable from node_modules (tsconfig paths don't
apply to node_modules). PR #145 originally had a 'Download libraries dist
artifact' step plus a script that rewrote package.json to depend on the
downloaded dist via file: links, ensuring node_modules/@valtimo/* exists
with proper main/module entries. The next-minor merge silently dropped
both steps; restoring them.

Klaas-Ritense committed:
cicd: quote the step name with a colon in it

YAML treated the unquoted colon in 'Rewrite local @valtimo/* deps to file:
links and install' as a key-value separator, breaking the workflow file
and causing startup_failure for the run that called it.

Klaas-Ritense committed:
cicd: strip workspaces field before file: dep rewrite

The next-minor merge added 'workspaces: [projects/valtimo/*]' to
package.json. With workspaces declared, npm install symlinks
node_modules/@valtimo/* to source folders (no compiled output) and
ignores the file:dist/valtimo/* deps the rewrite step writes. The
plugin tarballs in node_modules then can't resolve @valtimo/* via
standard node resolution and esbuild bails.

Stripping the workspaces field lets npm honour the file: deps so
node_modules/@valtimo/* points at the pre-built dist. The app's
own source still resolves @valtimo/* via tsconfig.paths, so the
local dev/test build paths are unaffected.

Klaas-Ritense committed:
cicd: allow EUPL-1.2 in backend license check

The smartdocuments / freemarker / smtpmail plugin POMs declare their
license as the raw string 'Licensed under EUPL, Version 1.2 (the
"License");' (looks like the license header comment leaked into the
POM <license><name>), which wasn't on the allow-list and broke
checkLicense in the publish job.

- license-normalizer-bundle.json: add EUPL-1.2 bundle + two
  transformation rules so both the noisy 'Licensed under EUPL,
  Version 1.2 ...' form and a clean 'European Union Public Licence
  v1.2' form normalize to the canonical name.
- allowed-licenses.json: add 'European Union Public Licence,
  Version 1.2' to the allow-list. Valtimo itself is EUPL-1.2
  (see top-level LICENSE.md), so this matches the project's
  existing license policy.

Klaas-Ritense committed:
cicd: also allow EUPL-1.1 (policy accepts v1.1 and v1.2)

Per Valtimo licensing policy: EUPL v1.1 and v1.2 are accepted for
consumed OSS code (with static and dynamic linking allowed). Valtimo
core itself remains EUPL-1.2 and is not affected by this allow-list.

Klaas-Ritense committed:
cicd: bump smtpmail to 2.0.5 (BE+FE aligned)

bootRun on v13 broke against smtpmail 2.0.4. Upstream released 2.0.5
to fix it. npm @valtimo-plugins/smtpmail@2.0.5 declares Angular 19.2.20
peerDep (matches the project exactly); Maven com.ritense.valtimoplugins:
smtpmail:2.0.5 release in flight.

Klaas-Ritense committed:
cicd: dev/prod plugin split via fileReplacements

bootRun / npm start failed with 'Failed to resolve entry for package
@valtimo/plugin' because @valtimo-plugins/* tarballs in node_modules
import @valtimo/* from their compiled .mjs, and Vite resolves those
via standard node_modules lookup. With workspaces: ['projects/valtimo/*']
declared, npm install symlinks node_modules/@valtimo/<lib> to source
folders that have no main/module/exports, so Vite pre-bundling fails.
tsconfig.paths only applies to the project's own source, not to
node_modules.

Invert the existing dev-tools.ts <-> no-dev-tools.ts fileReplacements
pattern for plugins so dev never installs them:

- frontend/src/app/app-plugins.ts (new, dev stub): empty pluginImports
  and pluginSpecifications arrays.
- frontend/src/app/app-plugins.prod.ts (new, release): real plugin
  module imports + plugin specs (SmtpMail + DocumentGenerator +
  MailTemplate + TextTemplate).
- frontend/angular.json production config: add the file replacement.
- frontend/src/app/app.module.ts: collapse explicit plugin imports
  to a single import + spread.
- frontend/package.json: drop the @valtimo-plugins/* deps so npm
  install no longer pulls the tarballs.
- frontend/release-plugins.json (new): per-branch plugin pin file
  (v13: freemarker 8.3.6, smtpmail 2.0.5).
- .github/workflows/frontend_build_push_docker_image.yml: new step
  'Add release plugin dependencies' before the rewrite step. Reads
  release-plugins.json, npm pkg sets each entry, no-ops with a
  ::notice:: if the file is missing - safe for v12-stable cherry-pick.

Klaas-Ritense committed:
cicd: drop release/0.0.0 push trigger from publish_release.yml

The push-on-release/0.0.0 trigger was a development fixture so we
could exercise the umbrella from a feature branch without first
needing it on next-minor. End-to-end pipeline is now validated
(attempts 7-11) and the workflow is ready to ship via workflow_dispatch
only.

Also simplifies the publish_images expressions: with push gone, the
'github.event_name == push' fallback can drop and inputs.publish_images
is the single source.

Klaas-Ritense committed:
cicd: hold monorepo image publishing until parity (Option A)

The monorepo's backend/app/gzac and frontend/src/ are dev-loop
configurations - if shipped as ritense/gzac-{backend,frontend}:<v>,
they would not match the existing template-built production images.
Backend gap is a handful of resource files (closable with a
src/dev/resources <-> src/main/resources split). Frontend gap is
an entirely different Angular project. Both need to be closed in
follow-up PRs before this pipeline can safely publish those tags.

For PR #145, ship the libs-publish + GH-release parts of the
umbrella only. Jenkins remains the canonical source of the gzac
images via the existing template-repo build path; release.sh keeps
working unchanged.

- publish_release.yml: drop backend_release.build_and_push_image
  and frontend_release.publish_image jobs. Drop publish_images
  input.
- publish_release_candidate.yml: drop publish_images input.
- backend_publish_release.yml: drop build_and_push_image job +
  publish_images input.
- backend_publish_release_candidate.yml: drop publish_images input
  (was a no-op anyway - the RC workflow never invoked image build).
- frontend_publish_release.yml: drop publish_image job +
  publish_images input.
- frontend_publish_release_candidate.yml: drop publish_images input.

Sub-workflows backend_build_push_docker_image.yml and
frontend_build_push_docker_image.yml are kept untouched - they
remain in use by backend_ci.yml / frontend_ci.yml for PR test
images. Re-enabling image publishing in the release path will be
a one-job-block change per side once the parity gaps close.

Klaas-Ritense committed:
cicd: phase 1/2 - backend src/main split + frontend apps/ split

Attempts 14-16 in .claude/plans/cicd-release-pipeline-pr145.md.

Backend (Attempt 14):
- backend/app/gzac/src/main/ now carries the gzac-template
  production configuration: META-INF/spring/AutoConfiguration.imports,
  config/application.yml, config/pbac/all.role.json, processes.xml
  with scan=true, prod logback, GZAC banner, com.ritense.valtimo.Application.
- backend/app/gzac/src/dev/ holds the developer-loop config:
  application.yml + application-mysql.yml + hibernate.properties +
  dev logback + processes.xml scan=false + dev banner +
  GzacApplication.kt + the formViewModel sample loaders/handlers.
- backend/app/gzac/build.gradle adds a 'dev' source set that extends
  main, gated behind -PincludeDev=true (bootRun/bootRunWithDocker
  default to dev; bootJar/assemble/image build default to main-only).
- Backend WAR built with -PincludeDev=false produces an image
  structurally equivalent to ritense/gzac-backend:<v> (verified in
  Attempt 17 of the plan).

Frontend (Attempts 15-16):
- frontend/src/ moved to frontend/apps/dev/src/ (git mv preserves
  history on all showcase components, app-plugins.ts/.prod.ts, etc.).
- frontend/apps/gzac/src/ populated from the gzac-frontend-template
  with the dev/prod app-plugins split applied (app-plugins.ts empty,
  app-plugins.prod.ts importing @valtimo-plugins/{smtpmail,freemarker}
  for production builds via angular.json fileReplacements).
- frontend/apps/gzac/src/environments/environment.prod.ts carries
  production: true (template only had environment.ts).
- frontend/projects/valtimo/app-shell/ is a new ng-packagr library
  exporting CommonAppModule. CommonAppModule wraps the 44 feature
  modules shared by every Valtimo variant + 9 Formio register* calls
  in its constructor. Both apps/dev and apps/gzac AppModule shrink
  by ~45% by consuming it.
- frontend/conf/default.conf added (was missing from monorepo; copied
  verbatim from gzac-frontend-template; nginx proxy expects 'gzac-backend'
  upstream).
- frontend/Dockerfile takes ARG VARIANT=dev; COPY now reads from
  deployment/${VARIANT}/browser so the same Dockerfile packages
  whichever variant the release flow requests.
- frontend/angular.json: renamed valtimo-angular-console -> dev,
  added gzac project (outputPath deployment/gzac, fileReplacements
  pair for environment.prod + app-plugins.prod in production config),
  added @valtimo/app-shell library project.
- frontend/package.json: added build:dev/build:gzac scripts pinned
  to --configuration production; @bpmn-io/cm-theme dep + overrides
  for the feelers transitive.
- frontend/tsconfig.json: rebased @src/* to apps/dev, added
  @valtimo/app-shell path mapping + references to apps/{dev,gzac}/
  tsconfig.app.json + app-shell tsconfig.lib.json.
- .github/workflows/frontend_build_push_docker_image.yml: new
  'variant' input (default dev, release flows pass gzac), plumbed
  into 'npm run build:<variant>' and build-args VARIANT=<variant>
  on the docker build-push step.

Local verification (per Attempt 17 in the plan):
- ng build dev: 41 MB initial, 14 lazy chunks, succeeds.
- ng build gzac: 41 MB initial, 11 lazy chunks, succeeds.
- docker buildx build --build-arg VARIANT=gzac frontend/: succeeds.
- Backend assemble + docker buildx build: succeeds; image config
  byte-identical to ritense/gzac-backend:13.33.0 at the application.yml
  / processes.xml / pbac / banner.txt level.
- Both images boot locally against the dev compose; SPA renders,
  /api proxy reaches backend, Keycloak login flow initiates.

Klaas-Ritense committed:
cicd: re-enable image publishing in release path

Attempt 18 in .claude/plans/cicd-release-pipeline-pr145.md.

Inverse of commit 369d27dd70 ("hold monorepo image publishing
until parity (Option A)"), now that Attempt 17 confirmed the
new folders produce structurally equivalent images vs.
ritense/gzac-{backend,frontend}:13.33.0.

- publish_release.yml + publish_release_candidate.yml: add
  publish_images (bool, default true) input; plumb through to
  both backend and frontend per-side calls. v12-stable's
  release.sh will pass publish_images: false to keep its
  Jenkins-built image flow unchanged.
- backend_publish_release.yml: add publish_images input; add
  build_and_push_image job calling
  backend_build_push_docker_image.yml with
  push_to_dockerhub: true and release_image_tag: <project_version>,
  gated on inputs.publish_images. The publish job's needs:
  stays at [build_run_tests] only so libs can still publish
  independently when v12-stable disables the image job.
- backend_publish_release_candidate.yml: add publish_images
  input as a no-op (parity only - RC doesn't build images).
- frontend_publish_release.yml: add publish_images input,
  restore permissions.packages: write, add publish_image job
  calling frontend_build_push_docker_image.yml with
  variant: gzac (new line vs. pre-Option-A diff; without it
  the workflow defaults to dev and produces the dev console
  image), push_to_dockerhub: true,
  release_image_tag: <project_version>, gated on
  inputs.publish_images.
- frontend_publish_release_candidate.yml: add publish_images
  input as a no-op (parity only).

All six workflows parse cleanly with yq.

Klaas-Ritense committed:
docs: README.md for backend/app/gzac and frontend/apps/{dev,gzac}

Companion docs to Phase 1/2 of cicd/release-pipeline (commit
130d14aec3). Each README explains the variant's role
(template-equivalent prod vs. developer-loop), what content
is variant-specific vs. shared, and which Gradle / npm
invocations exercise it.

These were intended for the Phase 1/2 commit but missed the
staging pass.

Klaas-Ritense committed:
cicd: fix CI breakage post Phase 1/2 split

Three issues caught by the first release/0.0.0 dispatch
(GitHub Actions run 28098057744):

1. Backend :backend:app:gzac:sourcesJar failed with
   "Entry banner.txt is a duplicate but no duplicate
   handling strategy has been set". When -PincludeDev=true
   (gradle.properties default), src/main/resources/ and
   src/dev/resources/ both contribute banner.txt,
   logback-spring.xml, processes.xml, application.yml.
   The Phase 1/2 commit added duplicatesStrategy = INCLUDE
   to processResources but not to sourcesJar, which also
   walks both resource roots. Configure both.

2. Frontend lib karma tests failed with
   "Pattern .../frontend/src/assets/config.js does not
   match any file" and
   "Module not found: Can't resolve .../frontend/src/test.ts".
   The lib karma configs all require('../../../karma.conf.js')
   which pointed at src/assets/config.js -- now lives at
   apps/dev/src/assets/config.js after the apps/ split.
   The lib test targets in angular.json also had
   "main": "src/test.ts" pointing at the same workspace-root
   path. Rebase both onto apps/dev/.

3. frontend/package-lock.json was out of sync with
   package.json: package.json declares
   @bpmn-io/cm-theme@0.2.0 (added in Phase 1/2) but the
   lockfile still resolved 0.1.0-alpha.2. Resync via the
   npm install that happened during local libs build.

Klaas-Ritense committed:
cicd: rebase lib tsconfig.spec.json onto apps/dev/src/test.ts

Run 28098649282 / Frontend pipeline / build_and_test_libs / build_and_test
failed with:

  Error: /home/runner/work/valtimo/valtimo/frontend/apps/dev/src/test.ts
    is missing from the TypeScript compilation. Please make sure it is
    in your tsconfig via the 'files' or 'include' property.
  Error: File '/home/runner/work/valtimo/valtimo/frontend/src/test.ts'
    not found.

Each lib's tsconfig.spec.json had `"files": ["../../../src/test.ts"]`
which resolved to frontend/src/test.ts -- gone since Phase 1/2 moved
src/ to apps/dev/src/. The previous fix (8388b3d3b1) updated
angular.json's lib test "main" entries to apps/dev/src/test.ts but
the tsconfig.spec.json files were missed.

Rebase the 41 lib tsconfig.spec.json files to
"../../../apps/dev/src/test.ts".

(tsconfig.lib.json files have `"exclude": ["src/test.ts", ...]`
relative to each lib's own root; that's a no-op since libs don't
have a src/test.ts of their own, so leaving them alone.)

Klaas-Ritense committed:
Merge branch 'next-minor' into cicd/release-pipeline

# Conflicts:
#	backend/README.md

Klaas-Ritense committed:
refactor(frontend): inline CommonAppModule into each variant's app.module.ts

Removes the @valtimo/app-shell library introduced in Attempt 16 and
places the 51 shared feature modules directly into both
apps/dev/src/app/app.module.ts and apps/gzac/src/app/app.module.ts.
The 9 Formio register* / enableCustomFormioComponents constructor
calls move back to each variant's AppModule constructor (where they
lived before Attempt 16).

* frontend/apps/{dev,gzac}/src/app/app.module.ts: imports/exports
  the full feature-module list directly. Variant-specific bits
  (dev showcase modules, KlantinteractiesApi / OpenKlantTokenAuthentication
  for dev; SseModule for gzac; per-variant PLUGINS_TOKEN values)
  stay where they were.
* frontend/projects/valtimo/app-shell/: deleted.
* frontend/angular.json: drop the @valtimo/app-shell library project.
* frontend/tsconfig.json: drop the @valtimo/app-shell path mapping
  and the library's tsconfig.lib.json reference.
* frontend/{README,apps/dev/README,apps/gzac/README}.md: update
  prose to describe the inlined wiring.

Verified locally: `ng build dev` and `ng build gzac` both succeed.

Klaas-Ritense committed:
feat(backend): absorb valtimo-backend-template + evv demo into monorepo

Pulls two more template repos into backend/app/, completing the backend
half of the template-absorption work described in the plan.

* backend/app/valtimo/ (new module): src/main/ copied from
  valtimo-backend-template@v/13 verbatim. New build.gradle mirrors the
  monorepo gzac pattern but consumes
  :backend:dependencies:valtimo-dependencies (not the gzac aggregator)
  and the local-document-generation / local-resource / local-mail /
  milestones / object-management / objecten-api{,-authentication} /
  objecttypen-api / zaken-api modules — matches the template's
  build.gradle.kts line-for-line. Dockerfile re-uses the template's
  shape (eclipse-temurin:21-jre-noble, /app.jar entrypoint).
* backend/app/gzac/src/evenementenvergunning/: kotlin + resources
  copied from gzac-evenementenvergunning-demo-backend@v/13 verbatim.
  build.gradle gains a `-Pvariant=evenementenvergunning` switch that
  adds this overlay to sourceSets.main and excludes the gzac-template
  Application.kt (the evv variant carries its own that boots the
  demo DataService). Default variant=gzac is unchanged.
* settings.gradle: include :backend:app:valtimo.

Verified locally:
  ./gradlew :backend:app:valtimo:assemble                                    → valtimo-13.1.3.war
  ./gradlew -Pvariant=evenementenvergunning :backend:app:gzac:assemble       → gzac-13.1.3.war (with evv overlay)
  ./gradlew :backend:app:gzac:assemble                                       → unchanged gzac-template build

Klaas-Ritense committed:
feat(frontend): absorb valtimo-fe + evv-fe templates as new apps/ variants

Pulls valtimo-frontend-template@v/13 and gzac-evenementenvergunning-demo-frontend@v/13
into the monorepo as new Angular workspace apps, packaged into
ritense/valtimo-frontend and ritense/gzac-evenementenvergunning-frontend
respectively.

* frontend/apps/valtimo/: src copied verbatim from valtimo-frontend-template@v/13;
  src/tsconfig.{app,spec}.json rebased onto the apps/<v>/src layout
  (extends ../../../tsconfig.json, outDir ../../../out-tsc/{app,spec}-valtimo).
* frontend/apps/evenementenvergunning/: src copied verbatim from
  gzac-evenementenvergunning-demo-frontend@v/13. Same tsconfig depth bump.
  app.module.ts: replace the direct `@valtimo-plugins/freemarker` import
  with `pluginImports`/`pluginSpecifications` from a new `./app-plugins`
  pair, mirroring apps/gzac. Local builds stub the array empty; CI
  swaps in app-plugins.prod.ts via angular.json fileReplacements after
  installing the release-plugins.json pins.
* frontend/angular.json: clone the gzac project block twice for the new
  variants. Drop the gzac-only environment.prod replacement and (for
  valtimo) the app-plugins replacement, since those templates ship a
  single environment.ts. outputPath rebased to deployment/<variant>.
* frontend/tsconfig.json: add tsconfig references for both new variants.

Single frontend/Dockerfile already handles every variant via `--build-arg
VARIANT=<name>`; no per-variant Dockerfile needed.

Verified locally:
  ng build valtimo                 → deployment/valtimo (40.2 MB initial)
  ng build evenementenvergunning   → deployment/evenementenvergunning (similar)
  ng build dev / ng build gzac     → unchanged

Klaas-Ritense committed:
feat(ci): variant-aware release plumbing for backend + frontend

Plumb 'variant' through both image-build sub-workflows so the release
path publishes a Docker image per absorbed template, not just gzac.

Backend (backend_build_push_docker_image.yml):
- New 'variant' input (gzac | valtimo | evenementenvergunning, default gzac).
- New 'resolve_variant' job derives gradle module, extra Pvariant flag,
  image name, and build-type matrix per variant:
    * gzac -> :backend:app:gzac, no flag, gzac-backend, [production, test]
    * valtimo -> :backend:app:valtimo, no flag, valtimo-backend, [production]
    * evenementenvergunning -> :backend:app:gzac, -Pvariant=evenementenvergunning,
      gzac-evenementenvergunning-backend, [production]
- assemble + build_image jobs now reference the resolved module/path/image_name.
- Add docker.io push (was missing on backend) gated on push_to_dockerhub
  AND matrix.build_type == 'production'; tag is
  ritense/<image_name>:<release_image_tag>.

Frontend (frontend_build_push_docker_image.yml):
- Extend existing 'variant' input to accept valtimo + evenementenvergunning.
- New 'resolve_variant' job maps variant -> image_name (and on v12, also
  the per-variant Dockerfile path).
- ghcr.io image name + docker.io push tag now reference the resolved
  image_name. DOCKERHUB_IMAGE env var dropped.

Per-side release wrappers (backend_publish_release.yml,
frontend_publish_release.yml):
- Replace the single image-build call with three sibling jobs, one per
  release variant (gzac, valtimo, evenementenvergunning). Each is
  gated on inputs.publish_images so v12-stable opt-out still works.

Frontend package.json: add build:valtimo and build:evenementenvergunning
scripts so the image-build workflow's `npm run build:<variant>` step
resolves on the new variants.

After this commit, `release.sh <semver>` (or the umbrella's
workflow_dispatch) publishes six Docker Hub images per release:
  ritense/{gzac,valtimo,gzac-evenementenvergunning}-{backend,frontend}

Klaas-Ritense committed:
fix(ci): override working-directory for frontend resolve_variant job

resolve_variant doesn't check out the repo, but inherits the workflow-level
defaults.run.working-directory: frontend, so bash failed with 'No such file
or directory' on the frontend/ path before the case statement could run.

Klaas-Ritense committed:
refactor(backend): split evenementenvergunning into its own gradle module

Promotes the evenementenvergunning demo from a -Pvariant overlay on
:backend:app:gzac to a first-class :backend:app:evenementenvergunning
module. Matches the structure of :backend:app:valtimo and removes the
sourceset-collision dance (excluding gzac's Application.kt, etc.) from
gzac/build.gradle.

* backend/app/gzac/src/evenementenvergunning/ -> backend/app/evenementenvergunning/src/main/
  (preserves history via git mv).
* backend/app/evenementenvergunning/build.gradle (new): inherits the same
  gzac aggregator deps + smtpmail/freemarker/smartdocuments. Includes
  spring-boot-starter-data-jpa for DemoDataService and pins the main class
  to com.ritense.valtimo.ApplicationKt (the evv variant's Application that
  boots the sample-data loader).
* backend/app/evenementenvergunning/Dockerfile: same as gzac's, scoped to
  the new module.
* backend/app/gzac/build.gradle: strip the -Pvariant=evenementenvergunning
  block (sourceSet srcDirs + Application.kt exclude + dupe strategy +
  starter-data-jpa). Gzac module is back to plain dev/main split.
* settings.gradle: include :backend:app:evenementenvergunning.
* .github/workflows/backend_build_push_docker_image.yml resolve_variant:
  evenementenvergunning case now resolves to module=evenementenvergunning
  with no -Pvariant flag.

Verified locally:
  ./gradlew :backend:app:evenementenvergunning:assemble  -> evenementenvergunning-13.1.3.war
  ./gradlew :backend:app:gzac:assemble                   -> unchanged gzac build

Klaas-Ritense committed:
refactor(backend): dev-as-peer module split + frontend prod-mode parity

- backend/app -> backend/apps; extract gzac src/dev into backend/apps/dev
- uniform src/main-only release modules (gzac/valtimo/evv) + shared
  backend/gradle/valtimo-app.gradle convention (bootRunWithDocker, shared
  compose, publish-disable, single-war)
- standardize backend Dockerfiles (21-jre-noble, /app.war, one-file glob)
- CI: dev variant builds the PR test image; drop -PincludeDev + prod/test matrix
- frontend: environment.prod.ts per released variant so images run prod mode
- per-app READMEs, public demo .env.properties (+ gitignore), start scripts

Klaas-Ritense committed:
Merge next-minor into cicd/release-pipeline

# Conflicts:
#	frontend/package-lock.json

Klaas-Ritense committed:
ci: add dispatchable GHCR-only image test workflow

Manual workflow_dispatch that reuses the release image builders for the
chosen variant(s)/side(s) with push_to_dockerhub=false, so images go to
GHCR only (no Docker Hub / Maven Central / npm). For smoke-testing the
image build path and the resulting images.

Klaas-Ritense committed:
ci(temp): add push trigger to test_build_images for feature-branch run

Temporary: lets the workflow run from the feature branch without first
registering it on the default branch (workflow_dispatch needs the file on
default). Scoped to this file's own path; builds all 4 variants x both
sides on push. Revert once the workflow lands on the default branch.

Klaas-Ritense committed:
fix(ci): give frontend image builder a tag_kind to prevent dev/gzac collision

dev and gzac both publish the gzac-frontend image; the builder hardcoded a
-test tag for every variant, so within one run the dev and gzac legs
overwrote each other on gzac-frontend:<branch>-test.<run>. Mirror the backend
builder: resolve a tag_kind per variant (dev=test, gzac/valtimo/evv=production)
and use it in the generated tags. dev keeps -test (PR test-env path unchanged);
released gzac frontend now gets -production tags, distinct from dev.

Klaas-Ritense committed:
feat(ci): stamp project_version into backend + frontend images

Image builds didn't receive the release version, so images reported the
committed placeholders (backend gradle.properties 13.1.3, frontend
versions.ts 13.0.1) instead of the release/test version.

- backend image builder: add project_version input; pass -PprojectVersion
  to the assemble. Forwarded from backend_publish_release.
- frontend image builder: add project_version input; stamp it into
  projects/valtimo/shared/.../versions.ts (frontendLibraries) before the
  app build, since the app bundles that source via tsconfig paths.
  Forwarded from frontend_publish_release.
- test_build_images: add project_version input + an effective default
  (gradle.properties projectVersion + -test.<run>), passed to both image
  matrices so test images carry a verifiable version.

Klaas-Ritense committed:
fix(ci): locate versions.ts dynamically when stamping frontend version

Make the stamp step robust: grep for the versions.ts holding frontendLibraries
instead of hardcoding the path, and no-op safely if absent.

Klaas-Ritense committed:
ci: show stamped version in test_build_images run-name (re-trigger)

Klaas-Ritense committed:
fix(ci): version the frontend libs in test builds; skip stale lib cache on explicit version

v12's app bundles @valtimo/config from dist, so the frontend version must come
from the libs build, not the image-build source stamp. Pass project_version to
the libs job in test_build_images, and skip the content-only lib cache when a
version is given so dist is rebuilt with the requested version (otherwise a
cache hit serves a stale-versioned dist).

Klaas-Ritense committed:
chore(ci): drop temporary test-image push trigger and dead BRANCH_NAME_SAFE

test_build_images is now workflow_dispatch-only (the path-scoped push trigger was
only there to run it from the feature branch before it's on the default branch).
Also removed the dead BRANCH_NAME_SAFE step + redundant step-level BRANCH_NAME env
from the frontend image builder (matches v12).

Klaas-Ritense committed:
refactor(ci): remove dead publish_images switch (monorepo is sole image producer)

release.sh no longer triggers Jenkins to build the ritense/* images from the
template repos, so there is no scenario where a release should skip images; the
input was always true. Removed publish_images from the umbrella, per-side
release, and per-side RC workflows (it was an inert no-op in RC). Image jobs now
run unconditionally on release.

Klaas-Ritense committed:
Merge remote-tracking branch 'origin/next-minor' into cicd/release-pipeline

Klaas-Ritense committed:
ci(backend): keep .RELEASE suffix on published Maven artifacts

Klaas-Ritense committed:
Merge remote-tracking branch 'origin/next-minor' into cicd/release-pipeline
